climada/engine/impact.py

```diff
@@ -1111,6 +1111,76 @@ def _cen_return_imp(imp, freq, imp_th, return_periods):
 
         return imp_fit
 
+    def calc_imp_from_mat(self, imp_mat, freq):
+        """
+        Set Impact attributes from the impact matrix. Returns a copy.
+        Overwrites eai_exp, at_event, aai_agg, imp_mat.
+
+        Parameters
+        ----------
+        imp_mat : sparse.csr_matrix
+            matrix num_events x num_exp with impacts.
+        freq : np.array
+            array with the frequency per event
+        Returns
+        -------
+        imp : Impact
+            Copy of impact with eai_exp, at_event, aai_agg, imp_mat set.
+        """
+        eai_exp = self._eai_exp_from_mat(imp_mat, freq)
+        at_event = self._at_event_from_mat(imp_mat)
+        aai_agg = self._aai_agg_from_at_event(at_event, freq)
+        return eai_exp, at_event, aai_agg
+
+    def _eai_exp_from_mat(self, imp_mat, freq):
+        """
+        Compute impact for each exposures from the total impact matrix
+
+        Parameters
+        ----------
+        imp_mat : sparse.csr_matrix
+            matrix num_events x num_exp with impacts.
+        frequency : np.array
+            annual frequency of events
+        Returns
+        -------
+        eai_exp : np.array
+            expected annual impact for each exposure
+        """
+        freq_mat = freq.reshape(len(freq), 1)
+        return imp_mat.multiply(freq_mat).sum(axis=0).A1
+
+    def _at_event_from_mat(self, imp_mat):
+        """
+        Compute impact for each hazard event from the total impact matrix
+
+        Parameters
+        ----------
+        imp_mat : sparse.csr_matrix
+            matrix num_events x num_exp with impacts.
+        Returns
+        -------
+        at_event : np.array
+            impact for each hazard event
+        """
+        return np.squeeze(np.asarray(np.sum(imp_mat, axis=1)))
```

Comment:
So currently the `at_event` is calculated from an `impact` variable (line 1033) which considers potential deductibles, covers etc. The imp_mat does not previously account for those, right? Aka the at_event would have to be modified afterwards, if such conditions exist?

Reply:
(similary for eai_exp,  etc.)

Reply:
Yes, this is then to be added in this method, good point! My issue with the current implementation is that the `at_event` and `eai_exp` are computed in a very obscure manner inside of the `.calc` method.

Reply:
Best would probably be to have a class method `calc_impact` and `calc_residual_impact` (if with insurance stuff).

Reply:
If I am not mistaken, the deductible all act on the impact matrix as implemented right now? Thus, the `imp_mat` is actually not the risk matrix, but the residual risk matrix after insurance deductibles have been applied. 

My suggestion would thus be to disentangle that as it is unclear at the moment.